controller/gateway/controller_reconciler_utils.go

```diff
@@ -723,13 +723,26 @@ func countAttachedRoutesForGatewayListener(ctx context.Context, g *gwtypes.Gatew
 				)
 			}
 
-			count += int32(len(httpRoutes))
+			count += countAttachedHTTPRoutes(listener.Name, httpRoutes)
 		}
 	}
 
 	return count, nil
 }
 
+func countAttachedHTTPRoutes(listenerName gatewayv1.SectionName, httpRoutes []gatewayv1.HTTPRoute) int32 {
+	var count int32
+	for _, httpRoute := range httpRoutes {
+		if lo.ContainsBy(httpRoute.Spec.ParentRefs, func(item gatewayv1.ParentReference) bool {
+			return item.SectionName == nil || *item.SectionName == listenerName
+		}) {
+			count++
+		}
+	}
+
+	return count
```

Comment:
This looks like https://pkg.go.dev/github.com/samber/lo#readme-countby. 

WDYT?

Reply:
Actually, my implementation was flawed as we did not want to count the amount of matching `parentRefs`, but the number of routes instead. I changed it.